### DISCUSSION TITLE
App router, custom auth and cookies

### DISCUSSION THREAD

**Discussion by nicolasthy** [vercel/next.js#54879] [Help]
### Summary

Hello 👋 

I'm working on a project using the app router API and I'm having some issues while working with cookies.
Here's some context on how my app works:
- When user comes on the website, I try to retrieve the `accessToken` from the cookies
  - If there is a token, redirect to the dashboard
  - If none, redirect to the login
- When user arrives on the dashboard, I try to fetch all it's workspaces using the `accessToken` in the headers of the request and then set the `currentWorkspaceId` in the cookies using the first workspace retrieved
  - Then I try to fetch the users of this workspace using both the `accessToken` and the `currentWorkspaceId` in the headers of the request

This workflow is pretty common imo, so I believe others have already tackled this scenario.
My goal here is to work with server components for all of my pages and try to fetch the data as soon as possible.
What I'm trying to avoid is the:
- Navigate to /workspace/users
- Blank screen for a second
- Redirected to /auth/login because not logged in

Here's the code I have using the pages router API that does what I'm trying to achieve (for the auth):
```js
export const getServerSideProps: GetServerSideProps = async (context) => {
  const { query, req, res } = context;
  const accessToken = getCookie(COOKIES_KEYS.accessToken, { req, res });

  if (accessToken) {
    // Already logged in! Redirecting to dashboard...
  }

  if (query.external_id) {
    const { refresh_token } = await api.post(
      "/exchange", {}, { headers: { "X-Auth-external-ID": query.external_id as string }, context }
    );

    const { access_token } = await api.post(
      "/refresh", {}, { headers: { Authorization: `Bearer ${refresh_token}` }, context }
    );

    setCookie(COOKIES_KEYS.refreshToken, refresh_token, { req, res });
    setCookie(COOKIES_KEYS.accessToken, access_token, { req, res });

    return {
      // Logged in! Redirecting to dashboard...
    };
  }

  return { props: {} };
};
```

Now I tried to do the same using a server component in the app router, however:
- It's not possible to set cookies outside of a Server action or a Route handler
- Server actions are still in alpha and I don't want to use them
- Tried creating a Route handler to set the cookie but the cookie is never set in the browser and I can't seem to retrieve it

Does anyone know how to achieve this using the app router? The only solution I found was to fetch data and set cookies in client components, but that doesn't feel right because I no longer use server components which is a big feature from Next.js, isn't it?

**Comment by devjiwonchoi** [comment]
You can set cookies via route handlers, but only through returning a new Response with headers.

```ts
import { cookies } from 'next/headers'
 
export async function GET(request: Request) {
  const cookieStore = cookies()
  const token = cookieStore.get('token')
 
  return new Response('Hello, Next.js!', {
    status: 200,
    headers: { 'Set-Cookie': `token=${token.value}` },
  })
}
```

### Reference
- [Route Handlers Cookies](https://nextjs.org/docs/app/building-your-application/routing/route-handlers#cookies)

**Reply by nicolasthy** [reply]
Well maybe I'm doing something wrong but this does not work. Also there's something I clearly don't understand in this example from the docs, why do we get the token from the cookies to set the exact same value to the cookies in the response? 🤔 

Using the Route handler from the docs, I try to call the endpoint in my home page.tsx like this, but nothing is added to the cookies:
```js
export default async function Home() {
  await fetch("http://localhost:3000/api/cookies");

  return (
    <main className="flex h-screen w-full flex-col items-center justify-center bg-blue-100 px-3">
      ...
    </main>
  );
}
```

**Reply by devjiwonchoi** [reply]
It is because the `cookies` in the route handlers "get" the cookies sent from request only.

**Reply by nicolasthy** [reply]
Could you elaborate a bit more? I'm missing a brick here to fully understand what I'm doing wrong 😬 
A quick example of the flow that sets the cookie from the `page.tsx` if possible 🙏

**Reply by devjiwonchoi** [reply]
Could you share a repro if you don't mind? If you want to set cookies on `page.tsx` directly, you can use `cookies().set()` on Server Components also.

**Reply by nicolasthy** [reply]
Example is really simple, here's a sandbox: https://codesandbox.io/p/sandbox/github/codesandbox/sandbox-templates/tree/main/nextjs?file=%2Fapp%2Fpage.tsx%3A1%2C1

Also tried to do it using server actions but didn't seem to have them working on page load 😕

**Reply by devjiwonchoi** [reply]
Um can't find where you're messing up with cookies😅
